go/xorm/main.go

```diff
@@ -0,0 +1,275 @@
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/encrypt-query-language/go/goeql" // imported using local path until published
+
+	_ "github.com/jackc/pgx/stdlib" // PostgreSQL driver
+	"xorm.io/xorm"
+	"xorm.io/xorm/names"
+)
+
+// To setup postgres:
+// Run: docker compose up
+// To run examples
+// Run: go run .
+
+// Create a separate custom type for each field that is being encrypted, using the relevant go type.
+// This custom type can then be used to access the conversion interface to use toDB and fromDb.
+type EncryptedTextField string
+type EncryptedJsonbField map[string]interface{}
+type EncryptedIntField int
+type EncryptedBoolField bool
+
+// 2. Add to struct
+type Example struct {
+	Id                  int64               `xorm:"pk autoincr"`
+	NonEncryptedField   string              `xorm:"varchar(100)"`
+	EncryptedTextField  EncryptedTextField  `json:"encrypted_text_field" xorm:"jsonb 'encrypted_text_field'"`
+	EncryptedJsonbField EncryptedJsonbField `json:"encrypted_jsonb_field" xorm:"jsonb 'encrypted_jsonb_field'"`
+	EncryptedIntField   EncryptedIntField   `json:"encrypted_int_field" xorm:"jsonb 'encrypted_int_field'"`
+	EncryptedBoolField  EncryptedBoolField  `json:"encrypted_bool_field" xorm:"jsonb 'encrypted_bool_field'"`
+}
+
+func (Example) TableName() string {
+	return "examples"
+}
+
+// Use the conversion interface for a custom type.
+// Use goeql serialization/deserialization to convert data to json expected by CipherStash proxy, and convert
+// back to the original field type.
+//
+// When serializing, the table and column for that field needs to be passed.
+// So the json configuration has the correct table and column when passed to the proxy.
+//
+// This setup means that for each field that needs to be encrypted:
+// - a separate custom type needs to be created
+// eg `type EncryptedTextFieldOne string`
+//    `type EncryptedTextFieldTwo string`
+//
+// - ToDB and FromDB needs to be implemented for each custom type.
+
```

Comment:
There might be a better way to do this.

I wanted to ensure that insert statements didn't need to be changed and only plaintext values were retrieved. Using the conversion interface allows us to convert from the go app type (e.g string) to the underlying db type (jsonb).

To do that, I needed to create a custom type, to then call the ToDb/FromDb method on.

The serialization/deserialization functions have been extracted to the goeql package currently sitting locally in the same folder (naming to tbc), so they can all be reused.